Fix IsSubSetOfWithNullCheck (#11222)

mjolka opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11222

Regression? Last working version:

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->

IsSubSetOfWithNullCheck is supposed to return true when `other` is
a subset of `parent`.

This fixes an issue where restore can take a long time to run on
solutions with many references and packages.

A test solution was created with 42 projects, with project n referencing
projects 1..(n-1). Each project also had a couple of package references
and a NoWarn element.

Before this change, restore ran for 5 minutes, used ~14GB of memory,
and failed with an OutOfMemoryException ("Array dimensions exceeded
supported range.")

After this change, restore completed in under 5 seconds.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

mjolka committed:
Fix IsSubSetOfWithNullCheck (#11222)

IsSubSetOfWithNullCheck is supposed to return true when `other` is
a subset of `parent`.

This fixes an issue where restore can take a long time to run on
solutions with many references and packages.

A test solution was created with 42 projects, with project n referencing
projects 1..(n-1). Each project also had a couple of package references
and a NoWarn element.

Before this change, restore ran for 5 minutes, used ~14GB of memory,
and failed with an OutOfMemoryException ("Array dimensions exceeded
supported range.")

After this change, restore completed in under 5 seconds.

erdembayar commented:
@jeffkl 
Could please take a look at this PR? I believe you have context on this.

nkolev92 commented:
Can we please add tests for this?

mjolka committed:
Fix NodeWarningProperties.IsSubSetOf

`NodeWarningProperties.IsSubSetOf(other)` should return `true` when
`other.ProjectWide` is a subset of `ProjectWide` and each value of
`other.PackageSpecific` is a subset of the corresponding value in
`PackageSpecific`.

Note that the order is reversed compared to
`HashSet<T>.IsSubsetOf(other)`.

This commit fixes a couple of cases:

- when `PackageSpecific[a]` is a proper subset of `other.PackageSpecific[a]` (should return `false`)
- when `PackageSpecific[a]` is a proper superset of `other.PackageSpecific[a]` (should return `true`)
- when `PackageSpecific` is an empty dictionary and `other.PackageSpecific[a]` is empty (should return `true`)
- when `PackageSpecific` is `null` and `other.PackageSpecific[a]` is empty (should return `true`)

mjolka commented:
I've added unit tests and in the process found and fixed an issue with `NodeWarningProperties.IsSubSetOf`.

I've also created a repo with a solution I described in the original comment, where a restore runs for several minutes and dies with a OOM exception. You can find it here: https://github.com/mjolka/LotsOfDependencies

If you have any suggestions on how to turn that into a test case, please let me know.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/TransitiveNoWarnUtilsTests.cs:
```diff
@@ -928,5 +929,184 @@ public void DependencyNodeEqualityFails_NodesHaveDifferentPackageIdsInPackageSpe
+        [InlineData("NU1605", null, true)]
+        [InlineData("NU1605", "", true)]
+        [InlineData("NU1605", "NU1605", true)]
+        [InlineData("NU1605", "NU1604", false)]
```
Can we add multiple log code test cases?
After all this was what the original change targeted.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/TransitiveNoWarnUtilsTests.cs:
```diff
@@ -928,5 +929,184 @@ public void DependencyNodeEqualityFails_NodesHaveDifferentPackageIdsInPackageSpe
+        }
+
+        [Fact]
+        public void NodeWarningPropertiesIsSubSetOfFails_PackageSpecificNoWarnsHaveDifferentKeySameValue()
```
naming: Some of these methods seem to be named differently. 

Can you please follow the recommended standard? https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#unit-test-method-naming
Thanks!

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/TransitiveNoWarnUtilsTests.cs:
```diff
@@ -928,5 +929,184 @@ public void DependencyNodeEqualityFails_NodesHaveDifferentPackageIdsInPackageSpe
+        }
+
+        [Fact]
+        public void NodeWarningPropertiesIsSubSetOfFails_PackageSpecificNoWarnsHaveSameKeyDifferentValue()
```
same here. 

Please use: `<what is being tested>_<under what conditions>_<with what expectations>`

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/TransitiveNoWarnUtilsTests.cs:
```diff
@@ -928,5 +929,184 @@ public void DependencyNodeEqualityFails_NodesHaveDifferentPackageIdsInPackageSpe
+            // Assert
+            result.Should().Be(expected);
+        }
+
```
Can we add a test case that assets similar log codes but different ids and there are with multiple ids in each.

nkolev92 reviewed (commented):
LGTM. 
Thanks for this change! 

Just some tweaks to the tests before approving.

erdembayar commented:
> I've added unit tests and in the process found and fixed an issue with `NodeWarningProperties.IsSubSetOf`.
> 
> I've also created a repo with a solution I described in the original comment, where a restore runs for several minutes and dies with a OOM exception. You can find it here: https://github.com/mjolka/LotsOfDependencies
> 
> If you have any suggestions on how to turn that into a test case, please let me know.

Thank you. Is performance improvement still holding after your recent changes?
Could you include before and after output windows log?

mjolka committed:
wip

mjolka committed:
wip

mjolka committed:
wip

mjolka committed:
wip

mjolka commented:
Sorry, pushed those changes by accident.

mjolka committed:
Replace PackageSpecificNoWarn with strings

nkolev92 reviewed (approved):
Awesome fix. 

Thanks @mjolka

nkolev92 commented:
@NuGet/nuget-client Last call for reviews. 
I'll try to merge Monday afternoon if there's no additional feedback.

mjolka commented:
@nkolev92 Thanks for your feedback. I was going to make some further changes to the test cases today but if you're happy with the state of the PR I'll leave it alone.

Would you like me to squash those "wip" commits and force push?

@erdembayar The performance improvement still holds after the recent changes.

I've attached the output of running `NuGet.exe restore .\src\LotsOfDependencies.sln -Force -ForceEvaluate` before and after the changes, is that what you were asking for?

[before.txt](https://github.com/NuGet/NuGet.Client/files/8189875/before.txt)

[after.txt](https://github.com/NuGet/NuGet.Client/files/8189868/after.txt)

nkolev92 commented:
@mjolka We squash and merge, no need to squash in your branch. I sometimes do it on the developing branch as it makes rebasing easier.

What kind of test changes were you thinking of?

mjolka commented:
@nkolev92 I was thinking of replacing the `PackageSpecificNoWarn` helper enum with a string based format for the test cases. Here's an example: https://github.com/mjolka/NuGet.Client/commit/41f8928d0f6d64cfdae06296e5207be0049e1b07

nkolev92 commented:
That's a good change. No changes to adding/removing test cases right?

mjolka commented:
@nkolev92 Right, no changes to adding/removing test cases. This would be the change: https://github.com/mjolka/NuGet.Client/commit/d4bd3e3f91806759a6c9198a3456f0dfc2b6e8c9

nkolev92 commented:
feel free to push if you'd like to. 
I'll rerun the tests when you do.

mjolka commented:
Thanks @nkolev92 .

I just noticed what might be a bug in `NodeWarningProperties.Intersect(HashSet<NuGetLogCode> first, HashSet<NuGetLogCode> second)` and wanted to get your thoughts on it as it's closely related to this change.

In the method above `Intersect`, `IsSubSetOfWithNullCheck`, `null` is considered to be the empty set:

https://github.com/NuGet/NuGet.Client/blob/aebfb49c190d4563cd157977928e3eeec7fe5212/src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs#L870-L890

Following that logic, `Intersect(first, null)` and `Intersect(null, second)` should both return `null` (or an empty `HashSet`). However, `Intersect(first, null)` returns `first` and `Intersect(null, second)` returns `second`.

https://github.com/NuGet/NuGet.Client/blob/aebfb49c190d4563cd157977928e3eeec7fe5212/src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs#L892-L913

nkolev92 commented:
I think it's ok if null and empty are considered the same for practical purposes. 
If you do want to make changes there, can we please do that in a different PR? :)
